fix exclude_border for the case of multidim input

orena1 opened the pull request:
## Description
When mulitdim exclude_border is supplied (to blob_log/blob_dog) it should be the same dimensions of the input (and this is checked in the first lines of _format_exclude_border). The returned exclude_border from  `_format_exclude_border` should have an additional dimension with trailing zero as peak_local_max https://github.com/scikit-image/scikit-image/blob/main/skimage/feature/blob.py#L536 gets image_cube which has an additional dimension (the last dim is not related to  the exclude_border variable). You can also see that when exclude_border is an int or not supplied, the return value is `(0,)* (img_ndim + 1) `


## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

lagru commented:
Hey @orena1, thank you very much for the patch! Do I understand you correctly?

`_exclude_borders` is supposed to ensure that the supplied input to the `exclude_borders` parameter is turned into a tuple whose length matches [`image_cube`](https://github.com/scikit-image/scikit-image/blob/main/skimage/feature/blob.py#L533) which has an additional dimension due to the stacking. E.g. this can be reproduced with 

```python
import numpy as np
from skimage import feature
feature.blob_dog(np.zeros((100, 100)), exclude_border=(3, 3))
# raises ValueError: `exclude_border` should have the same length as the dimensionality of the image.
```

In that case we should add a test that covers this case for `blob_log` and `blob_dog` in [test_blob.py](https://github.com/scikit-image/scikit-image/blob/main/skimage/feature/tests/test_blob.py). Would you mind adding that test?

orena1 commented:
Hi @lagru,
Yes `_format_exclude_border` should verify the `exclude_border` is correct, if it is an int/float it should create a tuple from it with the same dim on the input +1, when the last value of the tuple is 0. If it is a tuple it should only extend the tuple with a trailing zero.

I've added a test, and modified one test.
I hope it is ok.
Thanks.

lagru commented on skimage/feature/tests/test_blob.py:
```diff
@@ -155,19 +155,76 @@ def test_blob_dog_excl_border():
     )
     msg = "zero blobs should be detected, as only blob is 5 px from border"
     assert blobs.shape[0] == 0, msg
+    
+    # verify that tuple exclude_border work
+    blobs = blob_dog(
+        img,
+        min_sigma=1.5,
+        max_sigma=5,
+        sigma_ratio=1.2,
+        exclude_border=(exclude_border_dist,exclude_border_dist)
+    )
+
+    msg = "zero blobs should be detected, as only blob is 5 px from border"
+    assert blobs.shape[0] == 0, msg
```
I think this new snippet and the following new test `test_blob_dog_excl_border_multidim` are a bit redundant. I'd rather keep tests as short and contained as possible. How about consolidating this with `pytest.mark.parametrize` like this:
```python
@pytest.mark.parametrize("exclude_border", [6, (6, 6)])
def test_blob_dog_excl_border(exclude_border):
    # Testing exclude border
    # image where blob is 5 px from borders, radius 5
    img = np.ones((512, 512))
    xs, ys = disk((5, 5), 5)
    img[xs, ys] = 255
    # [...]
    blobs = blob_dog(
        img,
        min_sigma=1.5,
        max_sigma=5,
        sigma_ratio=1.2,
        exclude_border=exclude_border,
    )
    # [...]
```
Could you do the same for `test_blob_log_exclude_border` so that function is covered as well?

orena1 replied:
yeah, that is a good idea, I'm not very proficient in tests

lagru commented:
@orena1 thanks for the fast response! I've left a suggestion to make the test's a bit less repeating and to cover `blob_log` as well. Otherwise this looks almost ready to go.

orena1 commented:
Hi @lagru ,
I agree that your suggestion about test_blob_dog_excl_border, it does makes it much better.
But I don't think that the new test is redundant, it tests a different condition.
I assume I can do it all in one test,
Something like this:
```python
@pytest.mark.parametrize("disc_center", [(5,5), (5, 20)])
@pytest.mark.parametrize("exclude_border", [6, (6, 6), (4,20)])
def test_blob_dog_excl_border(disc_center, exclude_border):
    # Testing exclude border

    # image where blob is 5 px from borders, radius 5
    img = np.ones((512, 512))
    xs, ys = disk((5, 5), 5)
    img[xs, ys] = 255
    blobs = blob_dog(
        img,
        min_sigma=1.5,
        max_sigma=5,
        sigma_ratio=1.2,
    )
    assert blobs.shape[0] == 1, "one blob should have been detected"
    b = blobs[0]
    assert b[0] == disk_center[0] if type(disk_center) == int else disk_center[0],  <-- this part will become confusing, 
    assert b[1] == disk_center[1], "blob should be 5 px from x and y borders"
    
    blobs = blob_dog(
        img,
        min_sigma=1.5,
        max_sigma=5,
        sigma_ratio=1.2,
        exclude_border=exclude_border,
    )
    msg = "zero blobs should be detected, as only blob is 5 px from border"
    assert blobs.shape[0] == 0, msg
```

But I don't think it will be very readable.

How do you think I should procced  ?
Thanks

lagru commented:
> Something like this:

That suggestion seems totally fine to me except for this line:
```python
    assert b[0] == disk_center[0] if type(disk_center) == int else disk_center[0],  <-- this part will become confusing, 
```
I think it should be
```python
    assert b[0] == disk_center[0]
```
as `disk_center` is always a tuple of two integers if you look at your `parametrize` decorator.

orena1 commented:
Thanks for the comments, I think it does look much better. 
What do you think?

> as `disk_center` is always a tuple of two integers if you look at your `parametrize` decorator.

Yeah, sorry stupid mistake.

lagru commented on skimage/feature/tests/test_blob.py:
```diff
@@ -142,32 +142,42 @@ def test_blob_dog_3d_anisotropic(dtype, threshold_type):
-    msg = "zero blobs should be detected, as only blob is 5 px from border"
-    assert blobs.shape[0] == 0, msg
+    
+    if disc_center == (5, 20) and exclude_border == (4,15):
```
I prefer to keep test's really simple and not use conditionals to test different asserts but I'm fine to keep it in this case. If you want you can use a more general test that works with different `exclude_border` values. I think it should be 

```suggestion
    if np.all(np.array(disc_center) >= np.array(exclude_border)):
```

orena1 replied:
> I prefer to keep test's really simple and not use conditionals to test different asserts but I'm fine to keep it in this case.

Well, that is, in a way, why I thought it should be splitted to 2 functions. 

>If you want you can use a more general test that works with different exclude_border values. I think this should work

I thought about it, but I actually think that it is better to leave it like this, I think it is much easier to understand, as already from the values you can see the exact condition that it. Saves a comment I think.

Thanks

lagru commented on skimage/feature/tests/test_blob.py:
```diff
@@ -343,16 +354,24 @@ def test_blob_log_exclude_border():
-    msg = "zero blobs should be detected, as only blob is 5 px from border"
-    assert blobs.shape[0] == 0, msg
+    
+    if disc_center == (5, 20) and exclude_border == (4,15):
```
Same here...

grlee77 replied:
```suggestion
    if disc_center == (5, 20) and exclude_border == (4, 15):
```

orena1 commented:
Thanks for the comments @lagru , I've replied inline. What do you think?

lagru reviewed (approved):
Looks good on my part. Thanks!

grlee77 commented on skimage/feature/tests/test_blob.py:
```diff
@@ -142,32 +142,42 @@ def test_blob_dog_3d_anisotropic(dtype, threshold_type):
 
 
-def test_blob_dog_excl_border():
+@pytest.mark.parametrize("disc_center", [(5,5), (5, 20)])
+@pytest.mark.parametrize("exclude_border", [6, (6, 6), (4,15)])
```
```suggestion
@pytest.mark.parametrize("disc_center", [(5, 5), (5, 20)])
@pytest.mark.parametrize("exclude_border", [6, (6, 6), (4, 15)])
```

grlee77 commented on skimage/feature/tests/test_blob.py:
```diff
@@ -142,32 +142,42 @@ def test_blob_dog_3d_anisotropic(dtype, threshold_type):
-    msg = "zero blobs should be detected, as only blob is 5 px from border"
-    assert blobs.shape[0] == 0, msg
+    
+    if disc_center == (5, 20) and exclude_border == (4,15):
```
```suggestion
    if disc_center == (5, 20) and exclude_border == (4, 15):
```

grlee77 commented on skimage/feature/tests/test_blob.py:
```diff
@@ -329,11 +339,12 @@ def test_blob_log_3d_anisotropic():
-
-def test_blob_log_exclude_border():
-    # image where blob is 5 px from borders, radius 5
+@pytest.mark.parametrize("disc_center", [(5,5), (5, 20)])
+@pytest.mark.parametrize("exclude_border", [6, (6, 6), (4,15)])
```
```suggestion
@pytest.mark.parametrize("disc_center", [(5, 5), (5, 20)])
@pytest.mark.parametrize("exclude_border", [6, (6, 6), (4, 15)])
```

grlee77 reviewed (approved):
Thanks @orena1, this looks great. I will commit the few tiny pep8 suggestions I noticed and then we can merge

orena1 reviewed (commented):
Thanks
